Finishes rust docs for `channel_sv2`

lucasbalieiro opened the pull request:
closes #1810 

Add missing rust docs for:
- `protocols::v2::channels_sv2`
- `protocols::v2::channels_sv2::server`
- `protocols::v2::channels_sv2::client`
- removed `set_active_job` from trait `JobStore` and replace it with `add_active_job` where needed.

Addresses a few typos

lucasbalieiro commented:
Rebased with the changes from #1824 ;

Awaiting for #1826 and #1828 to be merged

plebhash commented on protocols/v2/channels-sv2/src/chain_tip.rs:
let's change a bit the description of `ChainTip` here

```diff
/// An abstraction over the chain tip, carrying information from `SetNewPrevHash` messages.
///
- /// Used while creating non-future jobs.
+ /// Used for:
+ /// - creating non-future jobs
+ /// - validating shares
```

plebhash commented on protocols/v2/channels-sv2/src/chain_tip.rs:
```diff
@@ -20,14 +21,17 @@ impl ChainTip {
         }
     }
 
+    /// Retrieves a clone of the hash of the previous block
```
no need to mention the fact that this is a clone

plebhash commented on protocols/v2/channels-sv2/src/lib.rs:
```diff
@@ -1,3 +1,30 @@
+//! - [`server`] — Mining server channel logic
+//! - [`template`] — Template and coinbase management
+//
+//! ## Usage
+//!
+//! Add `channels_sv2` as a dependency to your mining pool or SV2 proxy implementation to
+//! leverage robust channel state management and share validation logic.
+
```
I'd remove this paragraph

```suggestion
```

plebhash commented on protocols/v2/channels-sv2/src/server/error.rs:
```diff
@@ -1,30 +1,59 @@
+/// including missing chain tip, missing template, and job factory errors.
 #[derive(Debug)]
 pub enum GroupChannelError {
+    /// The chain tip is not set when required for the operation.
     ChainTipNotSet,
+    /// The specified template ID was not found among future jobs.
     TemplateIdNotFound,
+    /// Error occurred during job creation via the job factory.
     JobFactoryError(JobFactoryError),
```
these don't get rendered on Rust Docs, I'd remove them

plebhash commented on protocols/v2/channels-sv2/src/server/extended.rs:
```diff
@@ -1,4 +1,47 @@
+//!   the channel or network target and responds accordingly.
+//! - **Chain Tip Management**: Tracks the latest known chain tip (block height, previous hash,
+//!   timestamp, and target) for constructing headers and validating shares.
+//! - **Version Rolling**: Honors client configuration on whether version rolling is permitted,
+//!   validating submitted block versions accordingly.
```
```suggestion
//! - **Version Rolling**: Honors server configuration on whether version rolling is permitted,
//!   validating submitted BIP320 header versions accordingly.
```

this is server configuration, not client

yes, client can set `SetupConnection.REQUIRES_VERSION_ROLLING` but it's up to the server to allow that or not

plebhash commented on protocols/v2/channels-sv2/src/server/extended.rs:
```diff
@@ -1,4 +1,47 @@
+//! - Only one active job is allowed at a time. Jobs from a previous chain tip become stale when a
+//!   new chain tip is set.
+//! - Share acknowledgment logic is tied to a configured batch size (e.g., every `N` valid shares).
+//! - Extranonce validation supports dynamic updates but enforces consistency with previously agreed
+//!   parameters.
```
```suggestion
//! - Extranonce validation supports dynamic updates of `extranonce_prefix` but enforces consistency with previously agreed
//!   parameters.
```

plebhash commented on protocols/v2/channels-sv2/src/server/group.rs:
```diff
@@ -1,4 +1,36 @@
-//! Abstraction over the state of a Sv2 Group Channel, as seen by a Mining Server
+//! Sv2 Group Channel - Mining Server Abstraction.
+//!
+//! This module defines the [`GroupChannel`] struct, which provides an abstraction of a Stratum V2
+//! (SV2) group channel as maintained by a mining pool server.
+//!
+//! A group channel represents a logical grouping of standard channels, allowing multiple mining
+//! entities to share jobs and templates. It manages job distribution and activation for all
+//! associated standard channels, but delegates share validation and accounting to those standard
+//! channels.
```
```suggestion
//! This module defines the [`GroupChannel`] struct, which provides an abstraction of a Stratum V2
//! (SV2) group channel as maintained by a mining server.
//!
//! A group channel represents a logical grouping of standard channels, allowing multiple mining
//! entities to share jobs. It manages job distribution and activation for all
//! associated standard channels, but delegates share validation and accounting to those standard
//! channels.
```

plebhash commented on protocols/v2/channels-sv2/src/server/jobs/error.rs:
```diff
@@ -1,25 +1,51 @@
+/// including deserialization issues, invalid parameters, and protocol constraints.
 #[derive(Debug)]
 pub enum ExtendedJobError {
+    /// Failed to deserialize the coinbase transaction.
     FailedToDeserializeCoinbase,
+    /// Failed to deserialize coinbase outputs.
     FailedToDeserializeCoinbaseOutputs,
+    /// The coinbase transaction does not have the expected number of inputs.
     CoinbaseInputCountMismatch,
+    /// Failed to serialize coinbase outputs.
     FailedToSerializeCoinbaseOutputs,
+    /// Failed to serialize the coinbase prefix.
     FailedToSerializeCoinbasePrefix,
+    /// Attempted to convert a future job into a custom mining job (not allowed).
     FutureJobNotAllowed,
+    /// The minimum ntime provided is invalid.
     InvalidMinNTime,
```
these don't get rendered, they can be removed

plebhash commented on protocols/v2/channels-sv2/src/server/jobs/error.rs:
```diff
@@ -1,25 +1,51 @@
+///
+/// Represents possible failures when constructing or handling standard mining jobs.
 pub enum StandardJobError {
+    /// Failed to deserialize coinbase outputs.
     FailedToDeserializeCoinbaseOutputs,
```
this doesn't get rendered, it can be removed

plebhash commented on protocols/v2/channels-sv2/src/server/jobs/error.rs:
```diff
@@ -1,25 +1,51 @@
+/// including template errors, serialization issues, and invalid coinbase outputs.
 #[derive(Debug)]
 pub enum JobFactoryError {
+    /// The provided template is invalid. Includes a descriptive error message.
     InvalidTemplate(String),
+    /// Failed to deserialize coinbase outputs.
     DeserializeCoinbaseOutputsError,
+    /// Failed to serialize the coinbase transaction prefix.
     CoinbaseTxPrefixError,
+    /// Failed to serialize the coinbase transaction suffix.
     CoinbaseTxSuffixError,
+    /// Overflow occurred when summing coinbase outputs.
     CoinbaseOutputsSumOverflow,
+    /// The sum of coinbase outputs is invalid for the template.
     InvalidCoinbaseOutputsSum,
+    /// A chain tip is required but was not provided.
     ChainTipRequired,
```
these don't get rendered, they can be removed

plebhash commented on protocols/v2/channels-sv2/src/server/standard.rs:
```diff
@@ -199,30 +244,31 @@ impl<'a> StandardChannel<'a> {
 
         Ok(())
     }
-
+    /// Updates the current target for this channel.
     pub fn set_target(&mut self, target: Target) {
         self.target = target;
     }
-
+    /// Updates the nominal hashrate for this channel.
     pub fn set_nominal_hashrate(&mut self, nominal_hashrate: f32) {
         self.nominal_hashrate = nominal_hashrate;
     }
-
+    /// Returns the requested maximum target for this channel.
     pub fn get_requested_max_target(&self) -> &Target {
         &self.requested_max_target
     }
-
+    /// Returns the current target for this channel.
     pub fn get_target(&self) -> &Target {
         &self.target
     }
-
+    /// Returns the nominal hashrate for this channel.
     pub fn get_nominal_hashrate(&self) -> f32 {
         self.nominal_hashrate
     }
 
-    /// Updates the channel's nominal hashrate and target.
+    /// Updates the nominal hashrate and/or requested max target.
     ///
-    /// If requested_max_target is None, we use the cached value in the channel state.
+    /// Recalculates the channel's target difficulty.
+    /// Returns an error if the new target is invalid or exceeds the requested max target.
```
let's use this description instead

```
/// Updates channel configuration with a new nominal hashrate.
///
/// Adjusts target difficulty and internal state. Returns an error if
/// any input parameters are invalid or constraints are violated.
///
/// This can be used in two scenarios:
/// - Client sent `UpdateChannel` message, which contains a `requested_max_target` parameter that's also used as input
/// - vardiff algorithm estimated a new nominal hashrate, in which case `requested_max_target` is `None` and we use the value from the channel state (that was set either during channel opening or some previous `UpdateChannel` message)
```

plebhash commented on protocols/v2/channels-sv2/src/client/extended.rs:
```diff
@@ -1,4 +1,20 @@
+//! Extended Channels are primarily intended for use by mining proxies to enable more efficient
+//! management and distribution of the hashing space among downstream clients.
+//!
+//! ## Key Concepts
+//!
+//! An Extended Channel is characterized by:
+//!
+//! - **`extranonce_prefix`**: A prefix defined by the upstream pool, unique to each channel.
+//! - **`extranonce_size`**: The number of bytes available for use by the local proxy or
+//!   distribution to downstream clients for nonce rolling.
```
hmm it's characterized by more things?

but also, I guess we don't need a lengthy description on the module header
```suggestion
//! ## Key Concepts
//!
//! An Extended Channel is characterized by:
//!
//! - **`extranonce_prefix`**: A prefix defined by the upstream pool, unique to each channel.
//! - **`extranonce_size`**: The number of bytes available for use by the local proxy or
//!   distribution to downstream clients for nonce rolling.
```

lucasbalieiro commented:
@plebhash , thanks for the time taken in this review.
I've addressed the points you raised.

Rebased with the changes #1826;

Now awaiting for #1828.

lucasbalieiro commented:
Rebased with the changes from #1828.


Now awaiting for #1832

lucasbalieiro commented:
Rebase to get the changes from #1839 and added docs for the functions created in it.

Addressed the remaining points.
